File: 166-fraction-to-recurring-decimal/fraction-to-recurring-decimal.java
class Solution {
    public String fractionToDecimal(int numerator, int denominator) {
        if (numerator == 0)
            return "0";
        
        StringBuilder result = new StringBuilder();
        if (numerator < 0 ^ denominator < 0)
            result.append("-");        

        long n = Math.abs(Long.valueOf(numerator));
        long d = Math.abs(Long.valueOf(denominator));
        result.append(n / d);
        long rem = n % d;
        if (rem == 0) {
            return result.toString();
        }

        result.append(".");
        Map<Long, Integer> map = new HashMap<>();
        while (rem != 0) {
            if (map.containsKey(rem)) {
                result.insert(map.get(rem), "(");
                result.append(")");
                break;
            }
            
            map.put(rem, result.length());
            rem *= 10;
            result.append(rem / d);
            rem %= d;
        }

        return result.toString();
    }
}

File: 166-fraction-to-recurring-decimal/fraction-to-recurring-decimal.cpp
class Solution {
    using ll = long long;
public:
    string fractionToDecimal(int numerator, int denominator) {
        if (numerator == 0) return "0";

        bool ve = (numerator < 0) ^ (denominator < 0);
        ll n = abs<ll>(numerator * 1LL);
        ll d = abs<ll>(denominator * 1LL);
        if (d == 1) return ((ve) ? "-" : "") + to_string(n);
        cout << " I am here " << endl;
        if (!(n % d)) {
            return ((ve) ? "-" : "") + to_string(n / d);
        }

        string result = ((ve) ? "-" :  "") + to_string(n / d) + ".";
        ll rem = n % d;
        unordered_map<ll, int> seen;

        while (rem != 0) {
            if (seen.count(rem)) {
                result.insert(seen[rem], 1, '(');
                result.push_back(')');
                return result;
            }
            
            seen[rem] = result.size();
            rem *= 10;
            result += to_string(rem / d);
            rem %= d;
        }

        return result;
    }
};